homeassistant/components/nanoleaf/config_flow.py

```diff
@@ -0,0 +1,272 @@
+
+    # For discovery integration import
+    discovery_conf: dict
+    device_id: str
+
+    def __init__(self) -> None:
+        """Initialize a Nanoleaf flow."""
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle Nanoleaf flow initiated by the user."""
+        self.user_data_schema = (
+            ADVANCED_USER_SCHEMA if self.show_advanced_options else SIMPLE_USER_SCHEMA
+        )
+        if user_input is None:
+            return self.async_show_form(
+                step_id="user", data_schema=self.user_data_schema, last_step=False
+            )
+        self._async_abort_entries_match({CONF_HOST: user_input[CONF_HOST]})
+        self.nanoleaf = Nanoleaf(user_input[CONF_HOST])
+        try:
+            await self.hass.async_add_executor_job(self.nanoleaf.authorize)
+        except Unavailable:
+            _LOGGER.warning("Nanoleaf is not available at %s", self.nanoleaf.host)
+            return self.async_show_form(
+                step_id="user",
+                data_schema=self.user_data_schema,
+                errors={"base": "cannot_connect"},
+                last_step=False,
+            )
+        except NotAuthorizingNewTokens:
+            pass
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception("Unknown error connecting to Nanoleaf")
+            return self.async_show_form(
+                step_id="user",
+                data_schema=self.user_data_schema,
+                last_step=False,
+                errors={"base": "unknown"},
+            )
+        if user_input.get(CONF_MODE) == CONF_MANUAL:
+            return self.async_show_form(step_id="manual", data_schema=MANUAL_SCHEMA)
+        return self.async_show_form(step_id="link")
+
+    async def async_step_zeroconf(
+        self, discovery_info: DiscoveryInfoType
+    ) -> FlowResult:
+        """Handle Nanoleaf Zeroconf discovery."""
+        _LOGGER.debug("Zeroconf discovered: %s", discovery_info)
+        return await self._async_discovery_handler(discovery_info)
+
+    async def async_step_homekit(self, discovery_info: DiscoveryInfoType) -> FlowResult:
+        """Handle Nanoleaf Homekit discovery."""
+        _LOGGER.debug("Homekit discovered: %s", discovery_info)
+        return await self._async_discovery_handler(discovery_info)
+
+    async def _async_discovery_handler(
+        self, discovery_info: DiscoveryInfoType
+    ) -> FlowResult:
+        """Handle Nanoleaf discovery."""
+        host = discovery_info["host"]
+        name = discovery_info["name"].replace(f".{discovery_info['type']}", "")
+        await self.async_set_unique_id(name)
+        self._abort_if_unique_id_configured({CONF_HOST: host})
+        self.nanoleaf = Nanoleaf(host)
+
+        # Import from discovery integration
+        self.device_id = discovery_info["properties"]["id"]
+        self.discovery_conf = dict(load_json(self.hass.config.path(CONFIG_FILE)))
```

Comment:
why convert it to a dict?

Reply:
Also, this does I/O and should be run in executor.

Reply:
> why convert it to a dict?

`load_json` returns List or Dict, mypy expects Dict.
Is there a better way?

> Also, this does I/O and should be run in executor.

Fixed in https://github.com/home-assistant/core/pull/52199/commits/b2e5a517a65ed5f009258df37ef9d0f81e82ffd0

Reply:
Yes, cast it is what you want for typing: `cast(dict, value)`